lib/sycamore/sycamore/transforms/extract_graph_relationships.py

```diff
@@ -0,0 +1,187 @@
+                        "START_HASH": start_hash,
+                        "START_LABEL": relation["start_label"],
+                    }
+
+                    for key, value in relation.items():
+                        if key not in ["start", "end", "start_label", "end_label"]:
+                            rel["properties"][key] = value
+
+                    section["properties"]["nodes"][relation["end_label"]][end_hash]["relationships"][
+                        str(uuid.uuid4())
+                    ] = rel
+        return doc
+
+    def _serialize_relationships(self, entities):
+        from sycamore.utils.pickle_pydantic import safe_cloudpickle
+
+        serialized = []
+        for entity in entities:
+            serialized.append(safe_cloudpickle(entity))
+        return serialized
+
+    def _deserialize_relationships(self):
+        from sycamore.utils.pickle_pydantic import safe_cloudunpickle
+
+        deserialized = []
+        for entity in self.relationships:
+            deserialized.append(safe_cloudunpickle(entity))
+
+        return deserialized
+
+    async def _generate_relationships(self, section: HierarchicalDocument) -> dict:
+        relations = self._deserialize_relationships()
+        parsed_relations = []
+        parsed_metadata = dict()
+        parsed_nodes: dict[str, set] = defaultdict(lambda: set())
+        for relation in relations:
+            start_label = relation.__annotations__["start"].__name__
+            end_label = relation.__annotations__["end"].__name__
+
+            start_nodes = [
+                json.dumps(node["raw_entity"]) for node in section["properties"]["nodes"].get(start_label, {}).values()
+            ]
+            end_nodes = [
+                json.dumps(node["raw_entity"]) for node in section["properties"]["nodes"].get(end_label, {}).values()
+            ]
+
+            relation.__annotations__["start"] = Enum(start_label, {entity: entity for entity in start_nodes})
+            relation.__annotations__["end"] = Enum(end_label, {entity: entity for entity in end_nodes})
+
+            if start_nodes and end_nodes:
+                parsed_relations.append(relation)
+                parsed_metadata[relation.__name__] = {"start_label": start_label, "end_label": end_label}
+                parsed_nodes[start_label] |= set(start_nodes)
+                parsed_nodes[end_label] |= set(end_nodes)
+
+        if not parsed_relations:
+            return {}
+
+        # Use mypy ignore type since pydantic has bad interaction with mypy with creating class from a variable class
+        fields = {relation.__name__: (List[relation], ...) for relation in parsed_relations}  # type: ignore
+        relationships_model = create_model("relationships", __base__=BaseModel, **fields)  # type: ignore
+
+        entities = ""
+        for key, nodes in parsed_nodes.items():
+            entities += f"{key}:\n"
+            for node in nodes:
+                entities += f"{node}\n"
```

Comment:
O(n) more efficient to build up a list of strings and then join them once at the end (bc strings are immutable, so string += "hi" is a linear time op)